qiskit/transpiler/passes/optimization/_gate_extension.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Dynamically extend Gate classes with functions required for the Hoare
+optimizer, namely triviality- and post-conditionsto.
+A return value of 'true' for triviality conditions indicates the gate is
+always trivial, provided the qubit is in a classical state.
+Functions/gates that are omitted here are assumed to always be
+non-trivial and/or have unknown post-conditions.
+"""
+from z3 import Not, And
```

Comment:
Does this have to be in a separate module? If it does we should do a similar try except block around all the code here so we're not depending on z3 at import time.

Reply:
No I suppose it doesn't have to be, just thought it was cleaner that way. Let me know if I should move it to the main module.